### ISSUE TITLE
Create new default key bindings.

### ISSUE THREAD

**Issue by Viech** [Unvanquished/Unvanquished#544]
The old bindings are both outdated (e.g. item selection) and inefficient.

**Comment by lamefun** [comment]
The truth has finally been seen!

**Comment by Veyrdite** [comment]
Using 'x' for sprint is very awkward if you want to strafe (ie hold W and A) at the same time.  Moving this to the shift key make it much easier.

I remember you voicing issues about using the Y/Z keys before due to them being swapping in some keyboard layouts.  Do we want to officially support non-QWERTY layouts in the single set of default binds?

**Comment by Viech** [comment]
I think QWERTZ is quite common, though I don't know any numbers.

**Comment by mbasaglia** [comment]
If we can access the raw key id we could generate bindings depending on the physical key position instead if the associated character

**Comment by illwieckz** [comment]
it has to be done for next release.

this is a proposition:

```
[‾‾‾‾‾‾‾‾] [‾‾‾‾‾‾‾‾] [‾‾‾‾‾‾‾‾] [‾‾‾‾‾‾‾‾] [‾‾‾‾‾‾‾‾] [‾‾‾‾‾‾‾‾] [‾‾‾‾‾‾‾‾]
[ ~      ] [ 1      ] [ 2      ] [ 3      ] [ 4      ] [ 5      ] [ 6      ]
[ consol ] [        ] [        ] [ barb   ] [        ] [        ] [        ]
[        ] [ blastr ] [ weapon ] [ nade   ] [        ] [        ] [        ]
[________] [________] [________] [________] [________] [________] [________]

[‾‾‾‾‾‾‾‾‾] [‾‾‾‾‾‾‾‾] [‾‾‾‾‾‾‾‾] [‾‾‾‾‾‾‾‾] [‾‾‾‾‾‾‾‾] [‾‾‾‾‾‾‾‾] [‾‾‾‾‾‾‾‾]
[ Tab     ] [ Q      ] [ W      ] [ E      ] [ R      ] [ T      ] [ Y      ]
[ score   ] [        ] [ forwd  ] [        ] [ evolve ] [ talk   ] [ ttalk  ]
[         ] [ medkit ] [        ] [ reload ] [ buy    ] [        ] [        ]
[_________] [________] [________] [________] [________] [________] [________]
 
[‾‾‾‾‾‾‾‾‾‾‾] [‾‾‾‾‾‾‾‾] [‾‾‾‾‾‾‾‾] [‾‾‾‾‾‾‾‾] [‾‾‾‾‾‾‾‾] [‾‾‾‾‾‾‾‾] [‾‾‾‾‾‾‾‾]
[ Caps      ] [ Q      ] [ S      ] [ D      ] [ F      ] [ G      ] [ H      ]
[           ] [ left   ] [ backwd ] [ right  ] [ mark   ] [ decon  ] [        ]
[           ] [        ] [        ] [        ] [        ] [        ] [        ]
[___________] [________] [________] [________] [________] [________] [________]
 
[‾‾‾‾‾‾‾‾] [‾‾‾‾‾‾‾‾] [‾‾‾‾‾‾‾‾] [‾‾‾‾‾‾‾‾] [‾‾‾‾‾‾‾‾] [‾‾‾‾‾‾‾‾] [‾‾‾‾‾‾‾‾]
[ Shift  ] [ ≷      ] [ Z      ] [ X      ] [ C      ] [ V      ] [ B      ]
[ wwalk  ] [ toggle ] [        ] [        ] [ taunt  ] [ beacon ] [ voice  ]
[ crouch ] [        ] [        ] [        ] [        ] [        ] [        ]
[________] [________] [________] [________] [________] [________] [________]
 
[‾‾‾‾‾‾‾‾‾‾] [‾‾‾‾‾‾‾‾‾‾] [‾‾‾‾‾‾‾‾‾‾] [‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾
[ Ctrl     ] [ Meta     ] [ Alt      ] [  Space                               
[ run      ] [          ] [ strafe   ] [  jump                                
[          ] [          ] [          ] [  jump/jetpack                        
|__________] [__________] [__________] [_______________________________________

U:   admin talk
P:   pause
F1:  vote yes
F2:  vote no
F3:  team vote yes
F4:  team vote no
F11: ScreenshotJPG
F12: ScreenshotPNG


  /‾‾‾‾|‾‾‾‾\
 /    [‾]    \
/     [ ]     \
|     [_]     |
|______|______|
|             |
|             |
|             |
|             |
|             |
\_____________/

left click:   attack
right click:  secondary attack
middle click: barb/grenade
scroll up:    previous weapon (blaster)
scroll down:  next weapon     (weapon)
```

I sometime think about moving alien `evolve` to the same key as human `medkit` so `activate` (`buy` in above drawing) may not only be used by humans in front of an armoury, but also by everyone whatever the team to interact with map stuff.

**Comment by slipher** [comment]
+1 for making evolve a different command from `activate`

**Comment by illwieckz** [comment]
OK, so I put the evolve action on the same key as medkit on the above screenshot.

About previous/next weapon we can already do this (this is what I do myself):

```
teambind 2 MWHEELDOWN "itemact weapon"
teambind 2 MWHEELUP "itemact blaster"
```

But it would be better to just make weapon cycling (`weapprev`/`weapnext`) autoswitching, it means grenade/medkit have to fire/use with `attack` instead of `useitem`.

It's important grenades (and perhaps medkit) are accessible through weapon cycling so beginners will find them, even if there is a dedicated bind for that: we must make beginners able to play with the lowest gap possible, then encourage them to learn to become pro-players. We may add an option to hide medkit/grenade from weapon cycle for pro-players.

We may remove entirely human medkit from cycling as players will learn to use it quickly if it's using the same key as alien evolve.

There is somewhere (where?) a work in progress code for grenade first player view, as seen in this screenshot:

[![grenade first player](https://www.unvanquished.net/wp-content/uploads/2017/02/nade.png)](https://www.unvanquished.net/wp-content/uploads/2017/02/nade.png)

Until this is implemented we would just have to display the related pictogram on screen.

**Comment by Viech** [comment]
> It's important grenades (and perhaps medkit) are accessible through weapon cycling so beginners will find them, even if there is a dedicated bind for that: we must make beginners able to play with the lowest gap possible, then encourage them to learn to become pro-players.

I'm borderline OK with this if and once grenade and medikit are things that you hold and fire for all players, as the screenshot above implies. But in the meantime you are not helping new players by first teaching a bad practice that might stick. I'd go as far and argue that being completely unaware of the presence of grenade or medikit is better than getting into the habit of scrolling through a list and selecting something in the middle of a fight, because in the former scenario you will eventually learn the keys while in the latter you may forever adopt a playstyle that is too slow to compete. Note also that one key for throwing a grenade (usuall "g") and one for healing is common practice in casual shooter games nowadays, every kid understands this, and our players will aswell. They are not _that_ incompetent.

As a side remark, I have been redesigning the jetpack with the potential to abolish weapon scrolling in mind. It was important to me that it didn't need to be selected, because that was really unfun in trem, especially if you also wanted to use grenades mid air…

In your proposal, you need to switch reload and buy. Reload has been on "r" for as long as FPS games exist, and "e" is a common "use" key. Otherwise it looks good!

**Comment by illwieckz** [comment]
In fact I use myself “e” for use and “r” for reload but it's just cargo cult and since reload occurs more often than use, it makes sense to exchange them, but I'm very ok to revert to _e for use and r for reload_ if that's as common as `asdf`, the “r” key is accessible enough.

For weapon cycling and strategy to have with newcomers, well, any game like quake3 or xonotic has weapon cycling *and* per-weapon bind. I like the idea of a game being accessible from the start, and the only thing preventing players to learn the pro way is laziness (and I'm lazy so it's ok :stuck_out_tongue: ).

I sill haven't learnt the 1 key for ckit toggle, I inferred there was one available but I only took time to discover it today by reading the `keybindings.cfg` file today (that was also because before @slipher's work on keyboard, all my num keys were probably not available, that's usually not the case in other idtech-derivated games).

About grenade, I know I spent months of tremulous gaming before finding the courage to bind it to a key I would remember. Everytime I use a bare profile for testing purpose I just avoid grenade because I never remembered the default bind key. It means other lazy people will just not use it, it's better if they at least use it by cycling. Cycling is a great way to discover weapons, and it's very common in FPS.

For medkit I'm very ok to entirely remove it from cycling. Players must learn the keybind from the start, it's vital (in all senses), grenade is only used when you have enough credit (hence enough skill), and feel confident enough with the game to buy it. To get enough credit and enough confidence people have to learn the medkit keybind first.

**Comment by Viech** [comment]
If the bug was fixed, then the "wallwalk toggles" options menu option should be enough, with the default of press-to-wallwalk because that's less confusing then suddenly walking on walls and having to figure out how to stop it again. Plus it's easier to control in fights as you said.

**Comment by illwieckz** [comment]
sure

**Comment by illwieckz** [comment]
reverted buy (activate) and reload to `e` and `r`.

**Comment by kentone** [comment]
"Shift" has been run since the ages of call of duty and crouch has been "CTRL" or "C"
The run one has been one of the keys that annoyed me the most when playing and is not set to shift!
+1 for G as grenade :)

**Comment by illwieckz** [comment]
Perhaps I'm an alien from another planet but I never found [C] for crouch as something, unless I use the index finger, which means I can't strafe right while running. It looks like the key was just chosen because of the C in the name… On the contrary the pinky can access both [Shift] and [<] keys. In Unvanquished, wallwalking may require a high dexterity, while running does not really require it: no one has to release “run” for 1/10s to achieve a movement. Wallwalking may leverage such dexterity.

Similarly, [R] for reload seems to have been chosen because of the R in the name, because the [E] key was more accessible. But this [R] key was good luck since it lefts both [Q] and [E] for games that has leaning functionality.

It may look better to use [E] to reload and [R] to evolve since Reload is more often use than Evolve, and Reload is typically done in action while Evolve is not, but R is accessible enough so there is not need to break user habits on this…

But well, run is only used by one team while both team requires a key to crouch or wallwak.

I've never played COD, but in any game I play since decades, the first think I do is to assign shift to crouch, because [C] is not usable. I would be curious to know on which key other people reassign crouch when they start a game for the first time.

**Comment by illwieckz** [comment]
Someone wrote on discord today:

> <clement_686> When i play a game with humans sometimes i might use the Medkit through accident it would by good if when i take the medkit i have the item in my hand and get a click for use it

I really don't like what does other games by making the player having a syringe in hand, but that's just I don't like _how it looks_. I still think it's not bad for newcomer to scroll their mouse scroll until they find the medkit and activate it with “Fire” action.

This is not making people learning bad habit, it's making the game accessible to occasional players, and that's OK to have occasional players. People who want to be better can learn to be better and use keybind. Hence the idea of an option to disable the medkit from weapon rotation.

**Comment by illwieckz** [comment]
> > It's important grenades (and perhaps medkit) are accessible through weapon cycling so beginners will find them, even if there is a dedicated bind for that: we must make beginners able to play with the lowest gap possible, then encourage them to learn to become pro-players.
> 
> I'm borderline OK with this if and once grenade and medikit are things that you hold and fire for all players, as the screenshot above implies. But in the meantime you are not helping new players by first teaching a bad practice that might stick.

I just thought about this: if we are borderline OK with direct weapon cycling because we are afraid it would not help new players by teaching bad practices that might stick, it means we act like current item selection does not exist at all.

The fact is direct weapon cycling is way faster and way more efficient than the current item selection mechanism. By being afraid of direct weapon cycling encouraging people to not use direct keys means we consider the only alternative of weapon cycling is direct keys, not the current item selection mechanism we have. This denial means this item selection feature is less good than no feature.

And in fact, I agree with that. I never felt the Tremulous item selection mechanism was usable _at all_. I'm in favor of deleting it.

**Comment by illwieckz** [comment]
So the idea:

human mouse roll: cycle weapon between primary and blaster
human weapon left click: fire
human weapon right click: weapon-specific option (mdriver zoom, overloading luci)
human weapon middle click: grenade

alien left click: bite/claws
alien right click: class-specific action (mantis speed, goon bounce, tyrant rage, marauder zap, etc.)
alien middle click: barb/spit

granger/ckit left click: build
granger/ckit right click when buildable selected from build menu: unselect
granger/ckit right click when no buildable selected: mark for deletion
granger/ckit middle click: deleted marked

**Comment by illwieckz** [comment]
WASD: move
Q: Action (interact with buildables like armoury for humans, and in the future it may be usable for both aliens/humans to interact with map things)
E:  Reload (human weapon)
R: Evolve alien menu
F: Human health kit
1: Human primary weapon shortcut (alternative to mouse roll)
2: Human blaster shortcut (alternative to mouse roll)
3: Human grenade/Alien barb, spit (alternative to mouse middle click)
Shift: Human crouch/Alien wallwack
Ctrl or C: Human sprint
Alt: strafe
Space: jump/jetpack
V: Beacons
B: Voices
X: Taunt
Tab: Score
~: Console

**Comment by illwieckz** [comment]
spectator right click: team menu
spectator middle click: follow
spectator mouse roll: follow next

**Comment by illwieckz** [comment]
So we have to entirely remove the item selection mechanism, remove everything from weapon and blaster in selection list, and make `weapnext`/`weapprev` auto-activate.

**Comment by illwieckz** [comment]
@DolceTriade or @Viech, can you do the gamecode part of it?

**Comment by slipher** [comment]
What is the item selection mechanism? I don't even know what this is.

> remove everything except weapon and blaster in selection list, and make weapnext/weapprev auto-activate.

Then you would not even need those commands since they are equivalent to `itemtoggle blaster`.

**Comment by illwieckz** [comment]
> > remove everything except weapon and blaster in selection list, and make weapnext/weapprev auto-activate.
> 
> Then you would not even need those commands since they are equivalent to `itemtoggle blaster`.

Yes, although I would be in favor of keeping explicit `weapnext`/`weapprev` commands and make them alias to `itemtoggle blaster` for the early implementation so the binds would be already ready in user configuration file the day we implement first person grenade trhowing or things like that.

Note: I'm basically using that trick (itemtoggle on mouse roll) myself since ages now. But I consider this to be an hack I cannot put in default config. Putting this on default config will ask for trouble the day a mod implement a third weapon for example.

> What is the item selection mechanism? I don't even know what this is.

First, to be able to use it, you need to enable the tutorial mode because we never implemented currently selected item feedback on the item list.

Basically, while you roll your mouse, the item selection cycles between weapon, blaster, medkit, grenade (and previously, jetpack), then, when an item is selected (you can only know which item is selected by using tutorial mode), you press middle click to activate it: if it's weapon or blaster, it switches to it, if it's grenade, it throws it, if it's medkit, it uses it, and previously it started/stopped jetpack.

This is absolutely unneasy, bad and… broken. Even the tutorial indication is wrong. As long as an item is selected, activated or not, It says “Press MOUSE3 to use the <Something>”, for example: “Press MOUSE3 to use the <SMG>”. It displays this before you activate the SMG, and still displays this after you activated the SMG. There is no indication to roll the mouse to select the Blaster. So when you spawn as human, the tutorial says “Press MOUSE3 to use the SMG" and if you press MOUSE3 nothing happens! Because that's the weapon selected and it's already activated!

**Comment by slipher** [comment]
Oh, with the item selection mechanism you are talking about the current behavior of weapprev/next. Somehow I thought you were saying there is yet another way.

**Comment by illwieckz** [comment]
Yes, the current behaviour of weapprev/next. This is not usable in almost every aspect of it.

**Comment by slipher** [comment]
> Yes, the current behaviour of weapprev/next. This is not usable in almost every aspect of it.

Agreed. I've always thought that mechanism is far too slow to be used in this game, even for grenades. Like @kentone, I use g for grenades. I don't see a problem with using `itemtoggle blaster` directly in binds, since anything that involves selecting the grenade and activating it as two separate actions is too slow to me.

**Comment by illwieckz** [comment]
1. Using weapnext/weapprev commands is backward compatible: existing binds will work magically.
2. Using weapnext/weapprev commands is future proof: the binds will work the day someone wants to add a third weapon

Using `itemtoggle blaster` on mouse roll up/down is a hack to workaround the lack of the feature by relying on the fact we are lucky to currently only support two weapons. The hack will break as soon as someone adds a third weapon (we, or a modder). I'm not sure we need third weapon, but we already have 2 pistols modeled, we have that screenshot of first-person grenade etc. We cannot bet the future will never have more than 2 weapons.

I prefer to postpone the release to get this immediate weapon cycling on mouse roll in time, if that needs one week or something like that. This is really important.

**Comment by illwieckz** [comment]
I'm very OK to make `weapnext`/`weapprev` aliases of `itemtoggle blaster` at first as a quick&dirty implementation to get this it in time for the release.

**Comment by Gireen** [comment]
> granger/ckit middle click: deleted marked

that wont work with grangerspit
how about [ALT] [RIGHT CLICK] for deconstruct?

In general i think the binds should work like in most other shooters since it makes the first games much smoother. 
In that sens [R] should be reload cause that's where everyone expects it to be. 
[Q] Is common for communicating if i remember right. It where a good place for Beacons or Voice Commands to make them easier to use. But they would require the finished circle menu to work good.
[E] for interacting with things like map objects or armory and if nothing like that is around for evolve

[ENTER] for talk and [SHIFT] [ENTER] for ttalk additional to [T] and [Y]

Thats my opinion on the binds.  When i played for the first time after a while the weapon selection with 123 was a really positive experience. The weapon toggle with [1] feels just nice 😄 
Everything else looks ok to me. But talking about this is one thing, trying is another. 
I think it would  be good to make a bind file from this and just try them and see what works best.

**Comment by illwieckz** [comment]
> > granger/ckit middle click: deleted marked
> 
> that wont work with grangerspit

Oh right…

> how about [ALT] [RIGHT CLICK] for deconstruct?

I have two concerns with that:
- [low] it's hard to find by trial and error (aka exploration)
- [high] Alt is strafe, and the best use case for strafe is to move precisely the builder…

> In general i think the binds should work like in most other shooters since it makes the first games much smoother.

I agree, that's why I'm not against fro Crouch on C and Sprint on Shift if it's what does 90% of the games. The issue is that I change my keys in all games I play so I don't know what are the usual keys.

> In that sens [R] should be reload cause that's where everyone expects it to be.

The key layout I propose is not necessarily the one I use, because I assume my layout can be flawed. For example I use R myself to reload, but I proposed E because I did not found a good enough reason to do that (the fact Reload starts with R is not good enough if it's less reachable).

I just found the reason why R is good for reload, and why I configured R in all my games despite I find it stupid in Unvanquished: it keeps Q end E available for left and right leaning in games that supports leaning. So let's go for R for reload.

> [Q] Is common for communicating if i remember right. It where a good place for Beacons or Voice Commands to make them easier to use. But they would require the finished circle menu to work good.

Q for communicating would be weird, wasting such an accessible key is bad. I mean, if you gonna write a sentence or navigate a menu, you're going to interrupt your play. It's OK to use a less reachable key for that.

> [E] for interacting with things like map objects or armory and if nothing like that is around for evolve

The “if nothing like that is around“ condition is a good idea, it makes possible to keep the evolve menu and the buy menu on the same key, something that is really convenient.

My only concern is that the “if nothing like that is around“ condition is a bit vague. If we implement map interaction there would be cases where the evolve menu would pop-up at the wrong time, wasting precious timing and giving frustration.

Note that the medkit key is unused as alien so it can also be used to evolve as alien. It's just a bit less convenient to not use the same key to evolve as human and evolve as alien…

> [ENTER] for talk and [SHIFT] [ENTER] for ttalk additional to [T] and [Y]

Yep, such alternative is OK.

**Comment by illwieckz** [comment]
> > how about [ALT] [RIGHT CLICK] for deconstruct?
> 
> I have two concerns with that:
> 
>     * [low] it's hard to find by trial and error (aka exploration)
>     * [high] Alt is strafe, and the best use case for strafe is to move precisely the builder…

I have another strong concern about that, the `modcase` stuff does not work well (euphemism) with tutorial and bind option windows. Only direct key binds work correctly in all case.

**Comment by illwieckz** [comment]
This PR fixes almost all issues reported:

- https://github.com/Unvanquished/Unvanquished/pull/1182

It requires some other PR on engine and main package:

- https://github.com/DaemonEngine/Daemon/pull/330  
- https://github.com/UnvanquishedAssets/unvanquished_src.dpkdir/pull/15

**Comment by illwieckz** [comment]
To deconstruct what is marked for deconstruction, I suggest the `X` key, while being neighboring the WASD keys, this is not the most easy one, it's not a big loss to use this key for something that is not used intensively. Also the St. Andrew's cross shape of the `X` character can be a reminder for deletion.

**Comment by illwieckz** [comment]
I had the idea to use right mouse button to mark a buildable for deconstruction (like it is already to cancel a buildable), but that's not possible since right mouse button is used for the granger attack.

Or we would have to detect the buildable to make right mouse acting like a weapon if there is no buildable, or to mark for deconstruction if there is.

**Comment by illwieckz** [comment]
Those are the binds in my #1182 PR:

left mouse:
- fire weapon
- build menu for human ckit
- build menu for granger
- spectator open join menu

right mouse:
- mdriver zoom
- goon pounce
- marauder zap
- mantis dash
- tyrant trample
- granger swipe or mark for deconstruct if close to a buildable
- ckit mark for deconstruct if close to a buildable
- spectator quit follow mode

middle mouse:
- dragoon barb
- granger spit or deconstruct marked if close to a buildable
- spectator enter follow mode, cycle between first person follow and chase follow mod

Mouse wheel up:
- human previous weapon
- spectator follow previous player

Mouse wheel down:
- human next weapon
- spectator next previous player

WASD:
- move

Space:
- alien/human jump
- human fly with a jetpack
- specator fly up

Alt:
- speed (slow walk)

Shift:
- sprint

Ctrl:
- alien wallwalk
- human crouch
- spectator fly down

R:
- reload

Q:
- human activate armoury
- alien evolve

F:
- human medkit

G:
- human grenade



C:
- Taunt

B:
- Beacon menu

T:
- Talk

Y:
- Team talk

U:
- Admin talk

And some usual: F1 to vote yes, etc.

The tutorial was updated, it tells user how to chat, how to use jetpack, how to mark for deconstruct and deconstruct, etc.

**Comment by illwieckz** [comment]
Note: the true weapon cycling is implemented, both the mouse roll and the tutorial is already ready if someone wants to add a weapon.

**Comment by illwieckz** [comment]
Here is a video showcasing the deconstruct keybindings implemented in #1182 ([alternative non-youtube link](https://dl.illwieckz.net/b/unvanquished/wip/granger-attack3-deconstruction/20200427-165753-001.unvanquished-granger-contextual-mouse-behaviour.webm)):

[![Unvanquished granger contextual mouse behaviour](https://dl.illwieckz.net/b/unvanquished/wip/granger-attack3-deconstruction/20200427-165753-000.unvanquished-granger-contextual-mouse-behaviour.png)](https://www.youtube.com/watch?v=YDFi8Jo1SGc)

Behaviour of `attack2`/`attack3` is contextual. Usually, `attack2` is swipe and `attack3` is spit, but if you're close enough to deconstruct a friendly structure, `attack2` is mark for replacement and `attack3` is deconstruct marked. That's similar to the behaviour we already have when `attack2` cancels the build of a structure, even if that key is used to wipe if no structure is selected to swipe.

Note that on the very end I would like to have "simple right click” to mark/unmark for replacement, and to hold right click (with a visual feedback like the good pounce bar) a certain amount of time to deconstruct. This way we would free the middle button so people can bind grenade on middle mouse button without having to worrying when they play human. But that seems to be too much work to be done before the next release as we need a way to implement simple click and hold click on the same mouse button like we do with jump and jetpack.

At least this solution is elegant, is ready for next release, do not use complex bind (everything is on single key access) and is tutorial compliant (tutorial requires keybinds to be single key).

**Comment by kentone** [comment]
For the human armory keybind, is there a message that says something like: You have to be close to an armory to change your gear. (?) I remember that on tremulous took me a bit to understand how that worked.

Also I remember that was confusing what each construction was, I have to redownload the game and check how it is now, I see has some improvements :wink:

**Comment by illwieckz** [comment]
Good idea! We can display that if ammo is under a threshold, or if player spawned with the construction kit. We already have similar code to tell the player to return to medistation when he is low on health.

**Comment by illwieckz** [comment]
Fixed by  #1208. Yipee yay!
